Blog about integrating API Gateway with EventBridge and Lambda (#955)
pulumi/pulumi-hugo: themes/default/content/blog/api-gateway-to-eventbridge/index.md

---
title: "API Gateway to EventBridge with Pulumi"
date: 2022-01-07
draft: true
meta_desc: These are words that add up to at least 50 characters.

> meta_image: meta.png

Author: Yep, I'm on this now.

authors:
    - christian-nunciato
tags:
    - aws
    - api-gateway
    - eventbridge
    - lambda
---

If you're familiar with [Amazon API Gateway](https://aws.amazon.com/api-gateway/), you know it's all about making it easier to provision and manage a web API. Maybe you've used it, as I have, with [Crosswalk]({{< relref "/docs/guides/crosswalk/aws" >}}), our AWS extension library, to stand up a REST API and handle requests with AWS Lambda functions:

```typescript
import * as awsx from "@pulumi/awsx";

// Create a new API Gateway instance.
const api = new awsx.apigateway.API("my-api", {
    routes: [
        {
            // Define an HTTP endpoint.
            method: "GET",
            path: "/things",

            // Handle requests with an AWS Lambda function.
            eventHandler: async (apiGatewayEvent) => {
                return {
                    statusCode: 200,
                    body: JSON.stringify([
                        "thingOne",
                        "thingTwo",
                    ]),
                };
            },
        },
    ],
});

// Export the API's public URL. 🎉
export const apiUrl = api.url;
```

> I love this abstraction, and I use it all the time; it's an incredibly convenient way to solve a really common problem. But if you look closely at the code, you'll see that it's making a fairly strong assumption about how you'll be handling HTTP requests --- namely, that you'll do so with a single Lambda function, and that that function will always return a JavaScript object [of a particular shape](https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-integration-settings-integration-response.html).

Author: I must admit, I've always been a spaces-around-the-dashes kinda person. I've seen it both ways, and it seems to be one of those things that's left up to the publication to decide. I'm happy to standardize on no-spaces, though, but it'll be a hard habit to break. 😄

> I love this abstraction, and I use it all the time; it's an incredibly convenient way to solve a really common problem. But if you look closely at the code, you'll see that it's making a fairly strong assumption about how you'll be handling HTTP requests --- namely, that you'll do so with a single Lambda function, and that that function will always return a JavaScript object [of a particular shape](https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-integration-settings-integration-response.html).

Reviewer: I'm mostly going off of a number of style guides from my editing years, but for tech docs, I go with https://docs.microsoft.com/en-us/style-guide/punctuation/dashes-hyphens/emes

> I love this abstraction, and I use it all the time; it's an incredibly convenient way to solve a really common problem. But if you look closely at the code, you'll see that it's making a fairly strong assumption about how you'll be handling HTTP requests --- namely, that you'll do so with a single Lambda function, and that that function will always return a JavaScript object [of a particular shape](https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-integration-settings-integration-response.html).

Author: (I also personally like `---` (over `&mdash;`) 'cause quicker to type, generally works, and a bit easier on the eyes.)

> I love this abstraction, and I use it all the time; it's an incredibly convenient way to solve a really common problem. But if you look closely at the code, you'll see that it's making a fairly strong assumption about how you'll be handling HTTP requests --- namely, that you'll do so with a single Lambda function, and that that function will always return a JavaScript object [of a particular shape](https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-integration-settings-integration-response.html).

Reviewer: Yep! I personally don't have any preference (I've always used the HTML code, but that's really just a what-I-was-taught thing), so use whichever you'd like :) If you think it's easier for people to use/remember/read, it might be worth a style guide note.

> Indeed, this arrangement represents the API contract of a [Lambda _proxy_ integration](https://docs.aws.amazon.com/apigateway/latest/developerguide/http-api-develop-integrations-lambda.html) --- API Gateway integrations come in [many shapes and sizes](https://docs.aws.amazon.com/apigateway/latest/developerguide/http-api-develop-integrations.html); Lambda integrations just happen to be one of the more popular --- and much of the time, an approach like this one works out just fine. But depending on the needs of the application, it might not always be the best fit.

Author: I'll skip this one as I'm assuming it was a mistake (with the comma before the dash).

> Indeed, this arrangement represents the API contract of a [Lambda _proxy_ integration](https://docs.aws.amazon.com/apigateway/latest/developerguide/http-api-develop-integrations-lambda.html) --- API Gateway integrations come in [many shapes and sizes](https://docs.aws.amazon.com/apigateway/latest/developerguide/http-api-develop-integrations.html); Lambda integrations just happen to be one of the more popular --- and much of the time, an approach like this one works out just fine. But depending on the needs of the application, it might not always be the best fit.

Reviewer: Nope! If you're using am em dash and happen to need a comma (like a compound sentence as here), you place the comma before the final em dash as if the interjection is part of the previous clause. You can skip it if you want; I have a habit of editing grammar as I go :sheepish:

Imagine you were building a print-on-demand service, for example, and you wanted to expose an API to let your customers upload  documents and have them converted into orders. On AWS, you might reach for API Gateway, as above, to define an HTTP method and route (`POST /uploads`, say), wire it up to an AWS Lambda, and have the Lambda parse the upload, write the order to a database, and return a response. Visually, such a design might look something like this:

> ![Figure 1: Write new orders to the database](./figure-1.png)

Author: Good suggestion; I didn't realize these were so verbose! I'll revisit 'em, thanks for pointing it out.

> ![Figure 1: Write new orders to the database](./figure-1.png)

Reviewer: I'm by no means an expert on alt text; I just know it should be descriptive enough to convey the meaning of the image. I generally get decently descriptive on flowcharts like this because a lot of context is lost without it.

> ![Figure 1: Write new orders to the database](./figure-1.png)

Reviewer 2: heres a guide on writing alt text that id recommend: https://webaim.org/techniques/alttext/
specifically, id recommend removing the "The figure shows" language. Does the same information already appear in text form outside of the alt text? If so, it could potentially be reduced.

> ![Figure 1: Write new orders to the database](./figure-1.png)

Author: I've updated the alt text for each of these diagrams to express what they (focusing on content and context--thanks for sharing that doc @susanev).

It'd definitely work, and again, it's quite common. But at some point, you might find this tight coupling between API Gateway and Lambda too limiting. Say you wanted to be notified whenever a new order was received --- with a Slack message, maybe, in one of your team's shared workspace channels. Under the current design, you'd probably add a few lines of code to the Lambda function to import an HTTP library and make a call to the Slack API to post the message:

![Figure 2: Also notify Slack](./figure-2.png)

> That'd work, too --- but it'd be less than ideal for a number of reasons. For one, that Lambda would now have two jobs: taking orders and sending Slack notifications. That might be fine for today (it’s only a few lines of code, after all), but over time, those two jobs could easily become three, and then four, and soon enough, that poor Lambda could wind up becoming a lot more difficult to maintain. And given the importance of its main job --- capturing orders --- it's not something you'd want to risk failing at runtime because of a random Slack outage or other transient internet mishap. Moreover, with every bit of extra work you tack on to that function, you take a tiny step closer to hitting API Gateway's [30-second limit](https://docs.aws.amazon.com/apigateway/latest/developerguide/limits.html).

Author: I think I'd prefer to leave this one as-is, just 'cause it's a bit closer to how I actually talk, but if it's straight-up unclear, let me know and I'll come back to it.

> That'd work, too --- but it'd be less than ideal for a number of reasons. For one, that Lambda would now have two jobs: taking orders and sending Slack notifications. That might be fine for today (it’s only a few lines of code, after all), but over time, those two jobs could easily become three, and then four, and soon enough, that poor Lambda could wind up becoming a lot more difficult to maintain. And given the importance of its main job --- capturing orders --- it's not something you'd want to risk failing at runtime because of a random Slack outage or other transient internet mishap. Moreover, with every bit of extra work you tack on to that function, you take a tiny step closer to hitting API Gateway's [30-second limit](https://docs.aws.amazon.com/apigateway/latest/developerguide/limits.html).

Reviewer: Yep, up to you. My only concern is someone who might not have English as a first language may get lost in the longer sentences, which is why you'll generally find me attempting to break them up if it's feasible and readable.

Let's see how. Here's a revised architecture diagram showing how you might approach building an application like this one with EventBridge positioned between API Gateway and Lambda:

![Figure 3: Using EventBridge](./figure-3.png)

Now let's have a look at what it'd be like to build it with Pulumi. We won't build _everything_ in this diagram --- things like writing to the database or messaging Slack are left for you to explore --- but we will build enough to give you clear picture and a working example of how to connect all of these parts into a working application. Specifically:

* an API Gateway [instance](https://docs.aws.amazon.com/apigateway/latest/developerguide/http-api-vs-rest.html) to act as a container for your public API, along with a [_stage_](https://docs.aws.amazon.com/apigateway/latest/developerguide/http-api-stages.html) and a [_route_](https://docs.aws.amazon.com/apigateway/latest/developerguide/http-api-develop-routes.html) to handle inbound HTTP requests;

> * an EventBridge [integration](https://docs.aws.amazon.com/apigateway/latest/developerguide/http-api-develop-integrations.html) (comprised of an EventBus and EventRule) to handle notifications from API Gateway, and finally;

Author: Makes sense, yeah. I'll lowercase here, though, since in these bullets, I'm describing AWS resources more than the actual Pulumi ones (which come later, in the code).

* one or more Lambda functions to be invoked in response to event-rule matches.

Let's get started.

## Create a new project and stack

As always, it’s good practice to begin with a new project and stack:

{{< chooser language "typescript,python" />}}

{{% choosable language typescript %}}

```bash

> $ pulumi new aws-typescript

Author: Yep, I tried a variation with aws-native as well, but most of this stuff doesn't exist in the Cloud Control API yet, so we don't have it in aws-native, either.

```

{{% /choosable %}}

{{% choosable language python %}}

```bash
$ pulumi new aws-python
```

{{% /choosable %}}

Make sure you [configure your AWS credentials](https://www.pulumi.com/registry/packages/aws/installation-configuration) as well, and when prompted, choose whatever stack name and AWS region work best for you.

{{% notes %}}

In this post, we'll be using [API Gateway V2](https://docs.aws.amazon.com/apigatewayv2/latest/api-reference/api-reference.html), as it's newer and has few features (like [auto-deploy](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-apigatewayv2-stage.html#aws-resource-apigatewayv2-stage-properties)) that make it easier to work with in this context. The same ideas apply to both V1 and V2, though, so while V2 still lacks full feature parity with V1, the underlying resources we'll be using --- methods, routes, integrations --- are the same. You'll find links to examples of both at the end of this post.

{{% /notes %}}

## Create the API gateway and stage

{{% choosable language typescript %}}

> Start by replacing the contents of `index.ts` with the following code to declare a new API Gateway [API](https://www.pulumi.com/registry/packages/aws/api-docs/apigatewayv2/api/):

Reviewer: Out of curiosity, any reason you're using the `choosable` here instead of the `langfile`?

Reviewer: Also, I *think* the link will cause the build to break (it has in the past...):

```suggestion
Start by replacing the contents of `index.ts` with the following code to declare a new API Gateway [API]({{< relref "/registry/packages/aws/api-docs/apigatewayv2/api" >}}):
```

Author: Good callout, there are a few of these still left in here (side effect of writing in a different app); I'll fix 'em up.

Reviewer: I do the same thing :D

Author: > Out of curiosity, any reason you're using the choosable here

Only 'cause that's what I'm used to, but `langfile` would be nicer I agree. Thanks for spotting it!

    apiId: api.id,
    name: pulumi.getStack(),
    autoDeploy: true,
});
```

{{% /choosable %}}

{{% choosable language python %}}

```python
# Create a stage and set it to deploy automatically.
stage = aws.apigatewayv2.Stage("stage",
    api_id=api.id,
    name=pulumi.get_stack(),
    auto_deploy=True
);
```

{{% /choosable %}}

The next thing to do is register a route on the gateway to give your users a publicly accessible endpoint to upload to. But in order to do _that_, you'll need to tell API Gateway what to do when an upload happens. Since the plan of record is to notify EventBridge (using API Gateway’s built-in support for it), you’ll need to declare a few EventBridge things first.

## Add an event bus and event rule

Every AWS account gets an event bus by default (one that's aptly named `default`), but given how easy it is to create one, we might as well do that for this application. You'll also need to define an event rule --- a resource that watches a specific event bus for events that conform to a particular pattern or shape, and then routes those events to one or more targets (e.g., Lambda functions). Add the following lines to your program for both:

> ```typescript

Reviewer: So if I have Python on the chooser, this part stays as Typescript. Kinda switches up my brain a bit. Intentional?

Author: Nope, not intentional -- I'm working on adding all the Python stuff now as well. Should be there shortly.

...

// Create an event bus.
const bus = new aws.cloudwatch.EventBus("bus");

// Create an event rule to match
const rule = new aws.cloudwatch.EventRule("rule", {
    eventBusName: bus.name,

    // Specify an event pattern to watch for.
    eventPattern: JSON.stringify({
        source: ["my-event-source"],
    }),
});
```

The most noteworthy property of the `EventRule` resource is probably the `eventPattern`. EventBridge events all conform to a certain schema, and in this case, we’re expressing that this particular event rule should take action on any event that originates from `my-event-source`. (The `source` property is just a free-form string that by convention identifies the application or service responsible for the event.)

> With the event bus and event rule in place, you’re ready to define the integration itself — the resource responsible for connecting the gateway route (which we’ll get back to shortly) to your newly created event bus. As I mentioned earlier, there are several types of API Gateway integration to choose from, each one suited to a particular purpose. For this example, the `AWS_PROXY` type is good fit, as it’s simple and requires very little code; it doesn't give you quite as much control over the API Gateway response as you might like --- as a proxy, it just returns to the caller whatever is returned by the backend, in this case EventBridge --- but it's more than enough for the task at hand.

Reviewer: ```suggestion
With the event bus and event rule in place, you’re ready to define the integration itself, the resource responsible for connecting the gateway route (which we’ll get back to shortly) to your newly created event bus. As I mentioned earlier, there are several types of API Gateway integration to choose from, each one suited to a particular purpose. For this example, the `AWS_PROXY` type is good fit, as it’s simple and requires very little code; it doesn't give you quite as much control over the API Gateway response as you might like --- as a proxy, it just returns to the caller whatever is returned by the backend, in this case EventBridge --- but it's more than enough for the task at hand.
```

                    Effect: "Allow",
                    Principal: {
                        Service: "apigateway.amazonaws.com",
                    },
                },
            ],
        },
        managedPolicyArns: [
            "arn:aws:iam::aws:policy/AmazonEventBridgeFullAccess",
        ],
    },
);

// Create an API Gateway integration to forward requests to EventBridge.
const integration = new aws.apigatewayv2.Integration("integration", {
    apiId: api.id,

    // The integration type and subtype.
    integrationType: "AWS_PROXY",
    integrationSubtype: "EventBridge-PutEvents",
    credentialsArn: apiGatewayRole.arn,

    // The body of the request to be sent to EventBridge. Note the
    // event source matches pattern defined on the EventRule, and the
    // Detail expression, which just forwards the body of the original
    // API Gateway request (i.e., the uploaded document).
    requestParameters: {
        EventBusName: bus.name,
        Source: "my-event-source",
        DetailType: "my-detail-type",
        Detail: "$request.body",
    },
});

// Finally, define the route.
const route = new aws.apigatewayv2.Route("route", {
    apiId: api.id,

>     routeKey: "POST /uploads”,

Reviewer: ```suggestion
    routeKey: "POST /uploads",
```

    target: pulumi.interpolate`integrations/${integration.id}`,
});
```

With that, you’re ready to configure the Lambda.

## Add a Lambda function handler

> The hard part’s done: you’ve declared an API and route, mapped that route to an integration, configured the integration to publish events to an event bus, and defined an event rule to respond to those events. All that's left now is to tell the rule _how_ to respond.

Reviewer: ```suggestion
The hard part’s done: You’ve declared an API and route, mapped that route to an integration, configured the integration to publish events to an event bus, and defined an event rule to respond to those events. All that's left now is to tell the rule _how_ to respond.
```

Author: Keeping this one also since I believe this is technically okay to use either one.

So to finish things off, you need:

* a Lambda function to handle uploads,
* an EventBridge [_target_](https://docs.aws.amazon.com/eventbridge/latest/userguide/eb-targets.html) to bind your event rule to that function, and
* an IAM policy granting EventBridge permission to invoke the function.

Add the following lines to your program to complete it:

```typescript

> ...

Reviewer: Generally, I find some people add this, too, and get confused. Do you end up finding the same thing in TypeScript? I've been commenting that out (e.g., `// ...`) to ensure the code at least compiles or runs in an interpreter.

Author: Yeah, I like this better --- it's a bit tedious to have to remove them I agree. We could also build a bit more cleverness into the [copy-button script](the copy handler) to exclude them when they're all that's on the first line, as I don't think they're meaningful in any language. 🤔

{{< chooser language "typescript,python" />}}

{{% choosable language typescript %}}

```bash
$ pulumi new aws-typescript
```

{{% /choosable %}}

{{% choosable language python %}}

```bash
$ pulumi new aws-python
```

{{% /choosable %}}

Make sure you [configure your AWS credentials]({{< relref "/registry/packages/aws/installation-configuration" >}}) as well, and when prompted, choose whatever stack name and AWS region work best for you.

{{% notes %}}

In this post, we'll be using [API Gateway V2](https://docs.aws.amazon.com/apigatewayv2/latest/api-reference/api-reference.html) as it's newer and has a few features (like [auto-deploy](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-apigatewayv2-stage.html#aws-resource-apigatewayv2-stage-properties)) that make it easier to work with in this context. The same ideas apply to both V1 and V2, though, so while V2 still lacks full feature parity with V1, the underlying resources we'll be using---methods, routes, integrations---are the same. You'll find links to examples of both at the end of this post.

{{% /notes %}}

## Create the API gateway and stage

> Start by replacing the contents of {{% langfile %}} with the following code to declare a new API Gateway [API]({{< relref "/registry/packages/aws/api-docs/apigatewayv2/api" >}}):

Reviewer: Not sure why, but this link isn't pointing to the right spot in the preview (it's showing http://pulumi-hugo-origin-pr-955-01b72936.s3-website.us-west-2.amazonaws.com/blog/api-gateway-to-eventbridge/)
![Screen Shot 2022-01-11 at 4 17 33 PM](https://user-images.githubusercontent.com/1538692/149030375-bd6f58d3-edaf-4f59-b7fc-999bb493d3b1.png)

Author: Ah yeah, this is just a side-effect of not having the registry built into the preview. When we merge it, it'll point to https://pulumi.com/registry/packages/aws/api-docs/apigatewayv2/api. (There's a lil' note about this weirdness here: https://github.com/pulumi/pulumi-hugo#about-this-repository)

Reviewer: Huh; I've at least had the link render properly but 404 (or, at least, I think I have? Maybe that's been on local 😅 )

 +   ├─ aws:cloudwatch:EventBus       bus                       created
 ...

Outputs:
    apiURL: "https://geqfietbcl.execute-api.us-west-2.amazonaws.com/dev"

Resources:
    + 15 created

Duration: 31s
```

When the update finishes, you'll have a fully functioning API Gateway-EventBridge integration that you can verify with `curl`:

```bash
$ curl --data '{"some-key": "some-value"}' --header "Content-Type: application/json" \
     "$(pulumi stack output url)/uploads"

{"Entries":[{"EventId":"cdc44763-6976-286c-9378-7cce674dff81"}],"FailedEntryCount":0}
```

Note the response, which comes directly from EventBridge (courtesy of the `AWS_PROXY` integration), confirming the event was received and written to the event bus.

The final step is to confirm that the request made it all the way to Lambda, which you can easily do by tailing its output with `pulumi logs`:

```bash
$ pulumi logs --follow

Collecting logs for stack dev since 2022-01-06T16:18:48.000-08:00.
...

{
    source: 'my-event-source',
    detail: { 'some-key': 'some-value' }
}
```

When you're happy, be sure to tidy up with a `pulumi destroy`.

## What's next?

> There's a lot more you can do with integrations like this that didn't cover: add more Lambda function handlers to the program, have EventBridge target [other AWS services](https://docs.aws.amazon.com/eventbridge/latest/userguide/eb-targets.html) ([Step Functions](https://aws.amazon.com/step-functions) might be a good one to try next), validate HTTP request bodies (with API Gateway [models](https://docs.aws.amazon.com/apigateway/latest/developerguide/models-mappings.html), to keep bad data from ever reaching EventBridge), and more. Hopefully this gives you sense of what's possible, though. And as promised, you'll find examples using both API Gateway V1 and V2 at [pulumi/examples](https://github.com:pulumi/example/FIXME).

Reviewer: ```suggestion
There's a lot more you can do with integrations like this that didn't cover: add more Lambda function handlers to the program, have EventBridge target [other AWS services](https://docs.aws.amazon.com/eventbridge/latest/userguide/eb-targets.html) ([Step Functions](https://aws.amazon.com/step-functions) might be a good one to try next), validate HTTP request bodies (with API Gateway [models](https://docs.aws.amazon.com/apigateway/latest/developerguide/models-mappings.html), to keep bad data from ever reaching EventBridge), and more. Hopefully this gives you sense of what's possible, though. And as promised, you'll find examples using both API Gateway V1 and V2 at [pulumi/examples](https://github.com/pulumi/examples) on GitHub.
```

Author: Yep, I'll update this link when the examples are merged. Good eye!
